Implement platform pay payment method capturing for web

cornwe19 opened the pull request:
* Addresses #1122 
* Includes changes for detecting platform pay availability on web as well. (see #1478)

cornwe19 committed:
Add support for checking platform pay availability on web

* Partially addresses #1122

cornwe19 committed:
Stub in UI for platform pay button

* #1122

cornwe19 committed:
Fix formatting

*

cornwe19 committed:
Forward pay request options and type/style settings to web pay button

*

cornwe19 committed:
Hook up ability to capture payment methods via wallet pay on web

cornwe19 committed:
Fix formatting

*

cornwe19 committed:
Move payment request handlers formally ahead of the call to `show`

* Call to show is inside async block anyway, so logistically, nothing really changes here, but the intent of
   the code reads better with these declared before 'show'

*

cornwe19 commented:
@jonasbark @jamesblasco @remonh87 this supersedes #1478

cornwe19 commented:
@jonasbark @jamesblasco @remonh87 looking for feedback on this PR. It addresses a sizeable feature gap that exists on web.

remonh87 commented on packages/stripe_platform_interface/lib/src/stripe_platform_interface.dart:
```diff
@@ -99,6 +99,7 @@ abstract class StripePlatform extends PlatformInterface {
   /// Check if either google pay or apple pay  is supported on device.
   Future<bool> isPlatformPaySupported({
     IsGooglePaySupportedParams? params,
+    PlatformPayWebPaymentRequestCreateOptions? paymentRequestOptions,
```
I do not like the fact we need to inject a web specific method into the platform interface this should be independent.  Is this really required for just checking if platformpay is supported?

cornwe19 replied:
@remonh87 thoughts on my explanation above? Do you have suggestions for a different approach here given the constraints?

remonh87 replied:
no looks to me. I noticed some conflicts in this branch but we are good to go. @jonasbark  do you want to have a look at it as well?

cornwe19 replied:
Excellent. Thanks for your input. I'll get the conflicts cleared up shortly

remonh87 reviewed (commented):
Many thanks for this PR. I did review it and in general it looks good. However I am not convinced about adding the web paymentrequest options to isPlatformPaysupported interface. @jonasbark  what are your thoughts about it?

cornwe19 committed:
Fix return type for platformPayCreatePaymentMethod

*

cornwe19 committed:
Reformat code

*

cornwe19 commented on packages/stripe_web/lib/src/widgets/platform_pay_button.dart:
```diff
@@ -0,0 +1,115 @@
+  final BoxConstraints? constraints;
+  final ui.VoidCallback onPressed;
+
+  const WebPlatformPayButton(
```
Updated

cornwe19 commented:
@remonh87 I've resolved conflicts and reformatted code based on your suggestions. Let me know if there are any other outstanding items on this one. Excited to get this merged through

c.c. @jonasbark

remonh87 commented:
Thanks for the work. I am running the CI now and give Jonas till tomorrow to have a look at it else I will merge it.

cornwe19 commented:
@remonh87 Thanks!

I'm seeing some failures like `IntegrationError: Please call Stripe() with your publishable key. You used an empty string.` in stripe_js tests. That doesn't immediately feel like something I could have broken with this PR, but please let me know if there's some mitigation required here.

cornwe19 commented:
@remonh87 any remaining items blocking this pr?

cornwe19 commented:
@remonh87 Just synced this PR with latest main. I suspect checks will need to be rerun and will probably fail in similar ways to how they did on Jan, 24th. Please let me know if there's any action needed here. Hoping to get this PR merged soon.

Thanks!

jaxnz commented:
Excited to see the PR to be merged

cornwe19 committed:
Properly implement cancellation protocol when users dismiss pay sheet

*

*

cornwe19 commented:
@remonh87 FYI I just realized the way I had implemented payment request cancellation wasn't aligned with how iOS and Android implement it. Added a commit to update that (it now uses a `StripeException` with `LocalizedErrorMessage` having failure code `FailureCode.Canceled`)

Looking forward to getting this merged.

haomingzhang0101 commented:
Thanks for the great work! I am wondering when this feature will be released.
